src/protocol/serial.rs

```diff
@@ -0,0 +1,215 @@
+                stop_bits.unwrap_or(StopBits::DefaultStopBits),
+            )
+        }
+    }
+
+    /// Set the serial device's control bits. Per the UEFI specification, only five of the control
+    /// bits can be set with SetControlBits:
+    ///
+    /// * `REQUEST_TO_SEND`
+    /// * `DATA_TERMINAL_READY`
+    /// * `HARDWARE_LOOPBACK_ENABLE`
+    /// * `SOFTWARE_LOOPBACK_ENABLE`
+    /// * `HARDWARE_FLOW_CONTROL_ENABLE`
+    pub fn set_control_bits(&self, control: ControlBits) -> Result<(), Status> {
+        unsafe {
+            match (self.set_control_bits)(self, control.bits()) {
+                Status::Success => Ok(()),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Retrieve the serial device's control bits.
+    pub fn get_control_bits(&self) -> Result<ControlBits, Status> {
+        let mut bits: u32 = 0;
+        unsafe {
+            match (self.get_control_bits)(self, &mut bits) {
+                Status::Success => Ok(ControlBits::from_bits_truncate(bits)),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Write some data to the serial device.
+    pub fn write_raw(&self, data: *const u8, length: usize) -> Result<usize, Status> {
+        let mut new_length = length.clone();
+        unsafe {
+            match (self.write)(self, &mut new_length, data as *const CVoid) {
+                Status::Success => Ok(new_length),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Write a string to the serial device.
+    pub fn write(&self, data: &str) -> Result<usize, Status> {
+        self.write_raw(data.as_ptr(), data.len())
+    }
+
+    fn read_raw(&self, buf: *mut u8, length: usize) -> Result<Option<(*const u8, usize)>, Status> {
+        let mut new_length = length.clone();
+        unsafe {
+            match (self.read)(self, &mut new_length, buf as *mut CVoid) {
+                Status::Success => Ok(Some((buf, new_length))),
+                Status::Timeout => Ok(None),
+                e => Err(e),
+            }
+        }
+    }
+
+    /// Read `length` bytes from the serial device.
+    /// Note: The returned pointer is allocated with `allocate_pool`, and it is the caller's
+    /// responsibility to free at some point.
+    pub fn read_bytes(&self, length: usize) -> Result<Option<(*const u8, usize)>, Status> {
+        ::get_system_table()
+            .boot_services()
+            .allocate_pool(length)
+            .and_then(|buf_ptr| self.read_raw(buf_ptr, length))
+    }
+
+    /// Read a `length`-byte string from the serial device. The bytes will be interpreted as all
+    /// UTF-8 characters, since the serial device should only be returning 7-bit ASCII data.
+    pub fn read_str(&self, length: usize) -> Result<Option<&str>, Status> {
+        self.read_bytes(length).map(|option| {
+            option.map(|(bytes, read_length)| unsafe {
+                str::from_utf8_unchecked(slice::from_raw_parts(bytes, read_length))
```

Comment:
The conversion is in-place, so there's no leak. I think the best approach here might be to just expose a `read_bytes` which returns a slice of bytes instead, and let library consumers do with the bytes what they want.